src/backend/distributed/utils/resource_lock.c

```diff
@@ -332,7 +332,34 @@ LockShardDistributionMetadata(int64 shardId, LOCKMODE lockMode)
 	const bool sessionLock = false;
 	const bool dontWait = false;
 
-	SET_LOCKTAG_SHARD_METADATA_RESOURCE(tag, MyDatabaseId, shardId);
+	ShardInterval *shardInterval = LoadShardInterval(shardId);
+	Oid distributedTableId = shardInterval->relationId;
+	DistTableCacheEntry *distributedTable = DistributedTableCacheEntry(
+		distributedTableId);
+	uint32 colocationId = distributedTable->colocationId;
+
+	if (colocationId == INVALID_COLOCATION_ID ||
+		distributedTable->partitionMethod != DISTRIBUTE_BY_HASH)
+	{
+		SET_LOCKTAG_SHARD_METADATA_RESOURCE(tag, MyDatabaseId, shardId);
+	}
+	else
+	{
+		SET_LOCKTAG_COLOCATED_SHARDS_METADATA_RESOURCE(tag, MyDatabaseId, colocationId,
+													   shardInterval->shardIndex);
+
+		/*
+		 * NOTE: These next 2 lines are not really needed for correct locking.
+		 * However, we keep them so that we can have mixed cluster with nodes
+		 * that contain a Citus version earlier than 9.2. Those versions didn't
+		 * have the lock above, so the next two lines ensure mixed versions
+		 * still share a lock.
```

Comment:
Do we actually need this? lock_shard_metadata did not change its API.

Reply:
I guess not then, that makes sense. It will take different locks on different nodes (if you run mixed versions), but on the same node the same ones will be taken.